### PR TITLE
Add `fold()` operation as an inverse for `tf.image.extract_patches`

### LINKED ISSUE #52195: Support torch.nn.fold in tensorflow?
This function is popular in Vision Models. 

I don't find any api function in tf that is equivalent with [torch.nn.fold](https://pytorch.org/docs/stable/generated/torch.nn.Fold.html?highlight=fold#).

So does Tensowflow already have such a function or is there a plan to do so?

### PR DESCRIPTION
**Pull request by AbhishekChaudharii**
Fixes #52195 

## TL;DR
This PR introduces `tf.experimental.fold` operation which reconstructs images from patches extracted by `tf.image.extract_patches`. It supports configurable kernel size, stride, dilation, and *symmetric & VALID padding. Overlapping patch values are accumulated (summed) during reconstruction. `SAME` padding is not supported and raises `NotImplementedError`

This provides a native TensorFlow equivalent to PyTorch’s nn.Fold.

## Why is this needed?
TensorFlow has `tf.image.extract_patches` to extract patches but lacks the inverse operation to reconstruct images. This is a feature needed for computer vision workflows like vision transformers and patch-based models. The goal is to enable reconstruction of images from extracted patches.

## Implementation details
1. I have kept the implementation in `tensorflow/python/ops/array_ops.py` and exported as ` tf.experimental.fold`
2. For overlapping patches users should call `tf.config.experimental.enable_op_determinism()` to enable determinism before using `fold()` for reproducible results. Ref. https://www.tensorflow.org/api_docs/python/tf/scatter_nd.
3. Values of overlapping patches are accumulated (summed)
4. The current implementation only supports symmetric padding by passing an integer (eg. `padding=1`). `padding="SAME"` raises NotImplementedError
5. The API uses parameter names like `kernel_size` and `dilation` instead of `sizes` and `rates` like in `extract_patches` 
6. Followed NHWC format

## Work Completed
- Core implementation of `tf.experimental.fold()` in `array_ops.py`
- Added Comprehensive tests covering:
    - inverse property: fold(extract_patches(x)) = x (only applicable for non-overlapping cases)
    - Overlapping patches and non-overlapping patches
    - Compatible with multiple batch sizes and channels
    - Dilation and symmetric padding variations
    - PyTorch `nn.fold()` equivalence testing (locally, where applicable)
    - Gradient correctness

## Seeking feedback on
1. Module placement: Currently I have placed it `array_ops.py` exporting it to `tf.experimental`
    - Is this the correct location or another namespace would be more appropriate?
2. Parameter renaming: Should the parameter names align exactly with `tf.image.extract_patches`?

Thank you for reviewing this PR, happy to make follow-up changes based on feedback

### CODE REVIEW

**Comment by AbhishekChaudharii** [comment]
Hi @cantonios can you please review this PR when you get time? Would appreciate your feedback!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @AbhishekChaudharii,
  
  Thank you for proposing this valuable addition to TensorFlow! Implementing a native `fold` operation to complement `tf.image.extract_patches` has been a frequent request.
  
  Before this is ready for merging, we need to address a few API consistency and functionality points:
  
  1. **Parameter Naming:** 
     Currently, the API uses PyTorch-style parameter names (`kernel_size`, `dilation`). To stay consistent with the existing TensorFlow ecosystem, and specifically the forward operation `tf.image.extract_patches`, we should use TensorFlow convention:
     * Rename `kernel_size` to `sizes`
     * Rename `dilation` to `rates`
     
     Additionally, we might want to consider exposing `tf.experimental.unfold` as a symmetric alias to `tf.image.extract_patches` resolving any naming discrepancy.
  
  2. **Padding Support:**
     `SAME` padding is currently unsupported and raises a `NotImplementedError`. Because `SAME` padding is widely used in TensorFlow, we should add support for it before merging to ensure a complete implementation.
  
  3. **Overlap Reduction Options:**
     Currently, overlapping pixels are strictly accumulated (summed). In many reconstruction cases, users want to average the overlapping regions. We should either:
     * Provide a `reduction` parameter (e.g., admitting `"sum"` or `"mean"` configurations).
     * Provide documentation/example showing how users can normalize the summed output (e.g., dividing by a count-mask created by passing a tensor of ones through the same `fold` operation).
  
  4. **Determinism:**
     Please ensure we clearly document the behavior of overlapping accumulation and when users need to use `tf.config.experimental.enable_op_determinism()` to get reproducible results.
  
  Let us know your thoughts on implementing these changes, and we'd be happy to guide the PR through!

**Comment by AbhishekChaudharii** [comment]
Hi @dmiltr3, Thank you for the detailed feedback!
Kindly allow me some time to work through these points and update the PR accordingly. I will make the changes incrementally and post updates as I go. Thanks again for the review and guidance!

**Comment by AbhishekChaudharii** [comment]
Hi @dmiltr3, @cantonios,

I've addressed the review comments and pushed an update with the following changes:

1. **Parameters have been renamed** to stay consistent with `tf.image.extract_patches`
    - kernel_size to sizes
    - dilation to rates
    - stride to strides

2. **Support for `SAME` padding has been added** with the relevant tests.
    - Follows existing TF padding semantics, wherein for cases with odd number of total padding pixels, the extra padded pixels are added to the right and bottom sides of the tensor. 
    - Referred : "tensorflow/core/framework/kernel_shape_util.h" and "tensorflow/core/framework/kernel_shape_util.cc"

3. **Added reduction parameter** to give users options between summing and averaging overlapping pixels
    - To avoid zero division error, made a `divisor_matrix` where minimum value is at least 1, so that if any pixel is un-visited it won't have overlap count of 0. Thus avoiding element/0 which results in `inf` or 0/0 which results in `NaN` values.
    
4. **Updated the doc-string** with appropriate references and examples **to document the behavior of overlapping patches** and when to use `tf.config.experimental.enable_op_determinism()`

5. **Refactored the patch dimension validation logic to safely handle both static and dynamic shapes**.  The code can now run in graph mode also. 
    - Earlier when I ran tests with `@test_util.run_in_graph_and_eager_modes`  , `shape_internal` would be evaluated to a symbolic placeholder inside the `ValueError` string during graph tracing.

6. **Added a warning message** for when the user is performing the fold op. with overlapping patches(`stride<kernel` ) but hasn't enabled op determinism.


---

* **Question regarding determinism tests:** I added a `TestFoldDeterminism` test case where I explicitly disabled op determinism (`tf.config.experimental.disable_op_determinism()`). However, even when running the tests multiple times with larger input tensor sizes, the results remained deterministic.
  
Should we keep this test as-is, or is there a specific edge case or significantly larger input scale you'd recommend to reliably trigger the nondeterministic behavior of `scatter_nd` on the CPU/GPU?

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Here is a summary of the issues encountered with your PR and the fixes required to make it compatible with all TensorFlow execution modes (specifically Graph mode). Please apply these changes to your PR.

### 1. Graph Mode Compatibility in `array_ops.py`

#### A. Avoid Python `max()` with Tensors
In `array_ops.py`, calculating padding using Python's built-in `max()` fails when the inputs are symbolic Tensors (which happens when running in Graph mode or with dynamic shapes).

**Error:**
> `OperatorNotAllowedInGraphError: iterating over `tf.Tensor` is not allowed` (or similar depending on context)

**Fix:** Use `gen_math_ops.maximum()` or `tf.math.maximum()` instead.

```python
# Before
pad_total_h = max(0, (out_h - 1) * stride_h + k_eff_h - height)
pad_total_w = max(0, (out_w - 1) * stride_w + k_eff_w - width)

# After
val_h = (out_h - 1) * stride_h + k_eff_h - height
val_w = (out_w - 1) * stride_w + k_eff_w - width
pad_total_h = gen_math_ops.maximum(
    constant_op.constant(0, dtype=val_h.dtype), val_h
)
pad_total_w = gen_math_ops.maximum(
    constant_op.constant(0, dtype=val_w.dtype), val_w
)
```

#### B. Avoid Tensor evaluations in `if` statements
Checking if a Tensor is greater than zero directly in a Python `if` statement raises errors in Graph mode because the value is symbolic and not known at graph construction time.

**Problematic Code:**
```python
if pad_top > 0 or pad_bottom > 0 or pad_left > 0 or pad_right > 0:
    output = output[:, pad_top:pad_top+height, pad_left:pad_left+width, :]
```

**Fix:** Instead of checking padding values (which might be Tensors), branch based on the `padding` configuration argument or Python types.

```python
# Safer check for Graph mode
if padding == "SAME" or (isinstance(padding, int) and padding > 0):
    output = output[
        :, pad_top : pad_top + height, pad_left : pad_left + width, :
    ]
```

---

### 2. Test Session Handling in `array_ops_test.py`

When computing gradients using `gradient_checker_v2` in tests that might run in Graph mode, ensure a Session is available.

**Fix:** Wrap the gradient computation in `self.cached_session()`.

```python
# In TestFoldGradients.test_fold_gradient_numerical_correctness
def test_fold_gradient_numerical_correctness(self):
    # ... setup ...
    with self.cached_session():
        theoretical, numerical = gradient_checker_v2.compute_gradient(forward, [x])
    # ... assertions ...
```

---

Applying these fixes ensures that `tf.experimental.fold` works seamlessly in both Eager and Graph execution modes, and preserves the integrity of the test suite.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
# Feedback for PR #108937

We have analyzed your PR and suggested some improvements to make the implementation cleaner and more consistent with existing patterns.

## Issues Identified in Initial PR

1.  **Dead Code**: Replacing slashes before stripping them made `new_name.lstrip('/')` ineffective because there were no slashes left to strip.
2.  **Behavior Change on Leading Slashes**: Leading slashes were being preserved as underscores (e.g., `/tag` -> `_tag`) instead of being removed (e.g., `/tag` -> `tag`). This can break configurations that expect standard tag names.
3.  **Missing Warning**: Modifying `name` before checking `new_name != name` suppressed warnings for invalid characters (like multiple slashes).

## Recommended Implementation

Instead of explicitly replacing slashes in `clean_tag`, consider updating `_INVALID_TAG_CHARACTERS` to treat `/` as an invalid character. This integrates the replacement logic naturally.

### 1. Update `_INVALID_TAG_CHARACTERS`

Modify the regular expression in `summary_op_util.py` to remove `/` from the allowed set:

```python
# In summary_op_util.py

# Old: _INVALID_TAG_CHARACTERS = re.compile(r'[^-/\w\.]')
_INVALID_TAG_CHARACTERS = re.compile(r'[^-\w\.]')
```

### 2. Simplify `clean_tag`

Update `clean_tag` to strip leading slashes first, then use `_INVALID_TAG_CHARACTERS` to replace remaining slashes (and other illegal characters) with underscores.

```python
def clean_tag(name):
  # ... type checks ...

  if name is not None:
    new_name = name.lstrip('/')
    new_name = _INVALID_TAG_CHARACTERS.sub('_', new_name)
    if new_name != name:
      tf_logging.info('Summary name %s is illegal; using %s instead.' %
                      (name, new_name))
      name = new_name
  return name
```

### 3. Update Tests

This change flattens the hierarchy by replacing internal slashes with underscores. Existing tests that expect slashes to be preserved in OP names must be updated.

```python
# In summary_test.py
  def testSummaryNameConversion(self):
    # ...
    s3 = summary_lib.scalar('/name/with/leading/slash', c)
    # Old: self.assertEqual(s3.op.name, 'name/with/leading/slash')
    self.assertEqual(s3.op.name, 'name_with_leading_slash')
```

## Added Verification Tests

We have added unit tests to verify this behavior in `summary_ops_test.py`:

```python
  def testCleanTagWithSlashes(self):
    self.assertEqual('a_b', summary_op_util.clean_tag('a/b'))
    self.assertEqual('a_b_c', summary_op_util.clean_tag('a/b/c'))
    self.assertEqual('a_b_c_', summary_op_util.clean_tag('a/b/c/'))
    self.assertEqual('a_b', summary_op_util.clean_tag('/a/b'))

  def testCleanTagMultipleSlashes(self):
    self.assertEqual('a__b', summary_op_util.clean_tag('a//b'))
    self.assertEqual('', summary_op_util.clean_tag('///'))
```

**Comment by AbhishekChaudharii** [comment]
@dmiltr3 I think this feedback is for a different issue. My PR doesn't propose changes to `summary_op_util.py`

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
We encountered a failure in the API compatibility test (`api_compatibility_test`).

## Action Required
You added a new exported API `tf.experimental.fold` but did not update the API golden files. This is required for public API changes in TensorFlow.

Please run the following command in your environment to update the golden files and include them in your PR:

```bash
bazel run //tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True
```

This should update files under `tensorflow/tools/api/golden/`.

## Notes
- Ensure your local imports (if any) follow standard Open Source conventions (e.g., `from tensorflow.python...`).

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
### Minor Issues

1. **Logging Convention**: In `array_ops.py`, you used `tf_logging.get_logger().warning`. While valid, the standard convention in this file/package is often to use `tf_logging.warning` directly (if using `tf_logging`) or standard `absl.logging`. Using `tf_logging.warning` ensures that internal structured logging/bookkeeping is triggered.

2. **Robustness with Zero Channels**: In the `reduction='mean'` case, the code uses `gen_array_ops.ones_like(updates[:, 0])`. This assumes `updates` has at least one channel. If `patches` can have 0 channels, this will raise an IndexError. A safer alternative to get ones of shape `[TotalElements]` is `gen_array_ops.ones_like(indices[:, 0], dtype=updates.dtype)` or using `gen_array_ops.ones` with explicit shape, as `indices` is guaranteed to have the expected rank.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
## Really Minor Issues - Suggested Fixes

### 1. Missing Imports in `array_ops_test.py`
The new tests in `third_party/tensorflow/python/ops/array_ops_test.py` use libraries that are not imported at the top of the file. Please add the following imports:

```python
from tensorflow.python.framework import config
from tensorflow.python.framework import ops
from tensorflow.python.framework import random_seed
from tensorflow.python.ops import gradient_checker_v2
```

### 2. Local Imports in `array_ops.py`
In `array_ops.py`, inside the `fold` function (around line 6442), local imports are used:
```python
    from tensorflow.python.framework import config
    from tensorflow.python.platform import tf_logging
```
If these are intended to avoid circular imports, please ensure they are documented as such. Otherwise, consider moving them to top level if standard practices allow.

### 3. Formatting
Please run standard formatters (e.g., `yapf` or similar used in TF) to fix minor formatting issues.

**Comment by AbhishekChaudharii** [comment]
Thanks for the feedback!

**1. Missing Imports :** `tensorflow/python/ops/array_ops_test.py` already includes the required imports at the top.
The test for fold op exists in the above path and not in the `third_party/tensorflow/python/ops/array_ops_test.py`

**2. Local Imports:** I have already added an inline comment next to the local imports around line 6442, documenting that local imports are used specifically to avoid circular import issues.

3. Formatting: Local and Github Actions CI/CD check wrt `pylint` are passing. 
Referred: https://www.tensorflow.org/community/contribute/code_style

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
We have tested your PR internally. While the functionality seems sound, we noticed several minor style and formatting issues that should be addressed in GitHub before integration:

## Style & Formatting Issues

### 1. Missing Closing Bracket in Comments
In `tensorflow/python/ops/array_ops_test.py`, around line 345, the comment describing patches is missing a closing bracket for the outer list:
```python
    # patches = [[1, 2, 4, 5],
    #            [2, 3, 5, 6],
    #            [4, 5, 7, 8],
    #            [5, 6, 8, 9]  <-- Missing closing ]
```
It should be `[5, 6, 8, 9]]`.

### 2. Non-ASCII Character
In `tensorflow/python/ops/array_ops_test.py`, around line 368, a non-ASCII character `⊘` is used in the comment:
```python
    #* folded_img (with reduction='mean') = folded_img_summed ⊘ overlap count (ones_tensor)
```
Please use standard ASCII characters (e.g., `/` for division) to ensure compatibility.

### 3. Linter Findings (PEP 8 Violations)
There are numerous PEP 8 and style issues in the newly added code, particularly in `tensorflow/python/ops/array_ops.py` and `tensorflow/python/ops/array_ops_test.py`. We recommend running a linter (like `pylint` or `flake8`) locally to fix them. Common issues include:
- **Trailing whitespaces**: Many lines end with spaces.
- **Missing spaces after commas**: For example, `_range(0,batch_size,1)` should be `_range(0, batch_size, 1)`.
- **Wrong/Hanging Indentation**: Ensure consistent indentation throughout your changes.
- **Backslash line continuations**: Avoid using `\` for line continuation where possible (use parentheses instead).
- **Inconsistent quotes**: Try to match the quote style of the file.

Fixing these style issues will help speed up the review and integration process.

**Comment by AbhishekChaudharii** [comment]
Some places I have used this (`#*` )comment format eg. `#* folded_img (with reduction='mean') = folded_img_summed / overlap count (ones_tensor)`  for better visibility. Should I replace it with `#`?

**Review by dmiltr3 — DISMISSED** [review]

**Comment by AbhishekChaudharii** [comment]
Pushed a fix for formatting issue encountered in `ops/BUILD` and merged latest upstream. Ready for re-review when you have a moment

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
## Notes & Recommendations:

### 1. Argument Validation for `reduction`
We noticed that in `array_ops.py`, the `reduction` argument is documented to be either `'sum'` or `'mean'`, but there is no validation to enforce this. Currently, if an invalid value is passed (or anything other than `'mean'`), it silently defaults to `'sum'` behavior because `scatter_nd` accumulates by default.

Consider adding validation to raise a `ValueError` if `reduction` is not one of `('sum', 'mean')`:

```python
if reduction not in ('sum', 'mean'):
    raise ValueError(f"reduction must be 'sum' or 'mean', got {reduction}")
```

### 2. Dependencies
Ensure that if you are adding tests using `gradient_checker_v2` in `array_ops_test.py`, the corresponding build configuration (e.g., Bazel `BUILD` file if applicable) includes this dependency.

### 3. Unexecuted Test Cases
We noticed that in `array_ops_test.py`, the following functions are defined at the module level (with 0 indentation) but have `self` as an argument:
- `test_fold_same_padding_with_dilation`
- `test_fold_non_square_parameters`

Because they are at the module level, the test runner will not execute them. You probably intended to place them inside a `TestCase` class so they are actually run.

**Comment by AbhishekChaudharii** [comment]
The pre-submit code check test failed last time (post approval) was because of formatting issue in `BUILD`. gradient_checker_v2 dependency was already present but I think the formatter needs it sorted alphabetically.

Here's the snippet of log
```
not ok 1 Check buildifier formatting on BUILD files in 29ms
# `| xargs buildifier -v -mode=diff -diff_command="git diff --no-index"' failed with status 123
# buildifier formatting is recommended. Here are the suggested fixes:
# =============================
# --- a/tensorflow/python/ops/BUILD
# +++ b/tmp/buildifier-tmp-3411628562
# @@ -3996,9 +3996,9 @@ py_test(
#      deps = [
#          ":array_ops",
#          ":array_ops_gen",
# +        ":gradient_checker_v2",
#          ":math_ops",
#          ":random_ops",
# -        ":gradient_checker_v2",
#          "//tensorflow/core/config:flags_py",
#          "//tensorflow/python/eager:backprop",
#          "//tensorflow/python/eager:def_function",
```

I have applied the suggested fix. 

Input validation for reduction argument has been added with a relevant test.
Indentation for unexecuted test cases has been fixed. 

All CI checks should be passing now—ready for re-review/pull

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
We have run internal tests on your PR #108937 and they are **passing**. Your functional logic seems sound.

Here are some style suggestions and findings that could help improve the code quality and ensure a smooth merge:

### Style and Lint Findings

1.  **Line Length**: Some lines exceed the standard line length limit (typically 80 or 88 characters depending on the project setup). This is especially noticeable in some comments and imports.

    -   In `tensorflow/python/ops/array_ops.py`:
        ```python
        from tensorflow.python.framework import config #* Local imports - to avoid circular imports
        ```
        Consider breaking this line or placing the comment on the line above:
        ```python
        # Local imports - to avoid circular imports
        from tensorflow.python.framework import config
        ```

    -   In `tensorflow/python/ops/array_ops_test.py`:
        ```python
        # * folded_img (with reduction='mean') = folded_img_summed / overlap count (ones_tensor)
        ```
        Consider breaking long comments.

2.  **Docstring Punctuation**: One-line docstrings should end with a period `.` (or other appropriate punctuation).
    -   In `tensorflow/python/ops/array_ops_test.py`:
        ```python
        """Helper that matches TensorFlow's _extract_patches API"""
        ```
        Should be:
        ```python
        """Helper that matches TensorFlow's _extract_patches API."""
        ```

3.  **Local Imports**: The local imports inside functions are noted as intended to avoid circular imports. This is acceptable, but ensure they are necessary and documented as you did.

Please consider addressing these minor style issues.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Here are some style and lint issues that need to be addressed in GitHub:

## Style & Lint Issues

### 1. Docstring Punctuation & Formatting
Most of the new test cases in `tensorflow/python/ops/array_ops_test.py` have docstrings that are missing ending punctuation (periods) or have formatting issues.

Examples:
- `"""Helper that matches TensorFlow's _extract_patches API"""` -> Needs period.
- `"""Basic test of correctness for non-overlapping patches with VALID padding"""` -> Needs period.
- One-line docstrings should be followed by a blank line.

### 2. Line Length
Some lines exceed the standard line length limits (80 characters).

Examples:
- In `array_ops.py`, around line 6344 (in the `fold` implementation).
- Please verify all added lines and wrap them appropriately.

### 3. Hanging Indentation (Bad Continuation)
There are several instances of wrong hanging indentation in `array_ops.py`.

Examples:
- In error messages raised in `fold` implementation.
- In `scatter_nd` calls or other multi-line function calls.
- Please run a formatter or check indentation carefully.

### 4. Inconsistent Quotes
You have used inconsistent quote delimiters in both `array_ops.py` and `array_ops_test.py`. Please stick to the file's convention (usually single quotes `'`).

### 5. Imports Not at Top
In `array_ops.py`, there are imports not at the top of the file (around line 6467). If these are necessary to avoid circular dependencies, please add appropriate `pylint: disable` comments, otherwise move them to the top.

### 6. Unused Variables
There are unused variables (e.g., `i`) in `array_ops_test.py`. Please remove them.

## Code Review Feedback

### Functional Logic
> The check for overlapping patches should account for dilation. 
> Effective kernel size is `(kernel - 1) * dilation + 1`. 
> Patches overlap if `stride < k_eff`.
> 
> You already calculate `k_eff_h` and `k_eff_w` later. It might be better to move those calculations up and use them here.

**Comment by AbhishekChaudharii** [comment]
Thanks again for the review! I've pushed an updated branch that addresses all your feedback:

- Functional Logic: Moved the `k_eff` calculations up to compute the effective kernel size before the overlap check.

- Hanging Indentation: Fixed all the multi-line function calls, Applied a 4-space continuation indent strictly for hanging lines to differentiate them from TensorFlow's standard 2-space base indent.

- Line Limits: Replaced the hardcoded URL with symbol backticks (`tf.config.experimental.enable_op_determinism`) in the docstring to comply with the 80-character limit while ensuring it still renders as a link via tensorflow_docs.

- Style Constraints: Standardized to double quotes to match the surrounding file convention and resolved the missing docstring punctuation. Added a targeted # pylint: disable for the local circular import check.

**Review by dmiltr3 — APPROVED** [review]

**Comment by AbhishekChaudharii** [comment]
Hi @nithyak0204 @dmiltr3, `feedback/copybara` test was triggered right after the `awaiting review` tag was removed and it failed yesterday. Since all public CI checks have passed except `feedback/copybara`, could you please re-trigger that check or inspect the internal log and let me know if you need anything from my side? Thanks!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
### 1. Missing Release Notes in `RELEASE.md`
Because this PR introduces a new public API (`tf.experimental.fold`), please add a short summary of the new operation to `RELEASE.md` under the `Major Features and Improvements` or `TensorFlow API Changes` section. For example:
```markdown
* `tf.experimental.fold`: Added support for folding/reconstructing image tensors from extracted spatial patches as an inverse operation to `tf.image.extract_patches`.
```

### 2. Input Validation & Spatial Bounds Checking
In `tensorflow/python/ops/array_ops.py` (`fold` function):
- Validate that `output_size` contains positive dimensions:
  ```python
  if len(output_size) != 2 or output_size[0] <= 0 or output_size[1] <= 0:
    raise ValueError(f"output_size must be a tuple of 2 positive integers (height, width), got {output_size}")
  ```
- Add a spatial bounds check for `padding="VALID"` to ensure that the reconstructed patches fit within `output_size` before calling `scatter_nd`:
  ```python
  if padding == "VALID":
    if (out_h - 1) * stride_h + k_eff_h > height or (out_w - 1) * stride_w + k_eff_w > width:
      raise ValueError(
          f"output_size {output_size} is too small for extracted patches with sizes={sizes}, "
          f"strides={strides}, rates={rates}, and spatial patch shape ({out_h}, {out_w})"
      )
  ```
- Ensure safe dynamic rank checking in graph mode:
  ```python
  patches = ops.convert_to_tensor(patches)
  if patches.shape.ndims is not None and patches.shape.ndims != 4:
    raise ValueError(
        f"patches must be 4D (batch, height, width, patch_dim), got {patches.shape.ndims}D tensor with shape {patches.shape}"
    )
  ```

### 3. Test Isolation & Determinism Cleanup
In `tensorflow/python/ops/array_ops_test.py` (`TestFoldDeterminism`):
Calling `config.disable_op_determinism()` mutates global process state. Please ensure the prior op determinism state is saved in `setUp()` and restored in `tearDown()` / `addCleanup()` so subsequent tests in the suite are not affected:
```python
class TestFoldDeterminism(test.TestCase):

  def setUp(self):
    super().setUp()
    random_seed.set_seed(42)
    self._prev_determinism = config.is_op_determinism_enabled()

  def tearDown(self):
    if self._prev_determinism:
      config.enable_op_determinism()
    else:
      config.disable_op_determinism()
    super().tearDown()
```

### 4. Code Formatting & Cleanliness
- In `tensorflow/python/ops/array_ops.py`, fix indentation on the comment `# Get dimensions - extract dynamic shapes for the graph logic` to match the surrounding 2-space indentation.
- Remove redundant local import statements inside `fold` if they are already available at module scope, or keep them scoped at the top of the helper block if resolving circular dependencies.

**Review by dmiltr3 — APPROVED** [review]
Thank you for addressing all previous feedback points!

The updates look great:
* The release notes entry in `RELEASE.md` has been added.
* Input validation and spatial bounds checks for `output_size` and `patches` in `tf.experimental.fold` are well-structured.
* Test determinism flag state is properly preserved and restored in `TestFoldDeterminism`.
* Code style and comment indentation are clean.
* Internal presubmits and unit tests (`array_ops_test`, API compatibility golden checks) are all passing.

Approving this PR. Once the internal import process completes, this will be merged into the repository. Thanks again for your contribution!

**Comment by AbhishekChaudharii** [comment]
Thank you for the thorough reviews and guidance throughout the process. Really excited to see this feature land in TensorFlow!
